### PR TITLE
Plutus script purpose

### LINKED ISSUE #3533: Add ScriptPurpose to Conway
Conway requires two new ScriptPurposes:

* VotingPurpose (depends on #3542)
* ProposalPurpose (needed for runnable constitution, i.e. GovernancePolicy)

Related PR #3480

### LINKED ISSUE #3663: Implement redeemer pointers for new scripts
`Tag` in `RdmrPtr` need to add two new tags: one for voting another for proposing.

~The easiest way is to add two new constructors to the existing type and prevent new tags when decoding before conway with versioned decoder.~

A more type safe solution would be to convert `RdmrPtr` into a type family and add it to new `AlonzoEraScript` type class.

### PR DESCRIPTION
**Pull request by lehins**
# Description

Fixes #3663 
Fixes #3533 

Introduce PlutusPurpose:

ScriptPurpuse and RdmrPtr are very much related concepts. This commit
replaces them with a single type family that is parametric on the era
and on the type that does the selection whether it is a plutus script
purpose or it is a redeemer pointer. Those types are called `AsItem` and
`AsIndex` respectively.

Besides making plutus script purpose parametric this commit also
introduces new purposes for Conway: `Voting` and `Proposing`

# Checklist

- [ ] Commit sequence broadly makes sense and commits have useful messages
- [ ] New tests are added if needed and existing tests are updated
- [ ] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [ ] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [ ] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: eras/conway/impl/cddl-files/conway.cddl [hunk 1]
```diff
@@ -469,13 +469,19 @@ constr<a> =
   ; similarly for tag range: 6.1280 .. 6.1400 inclusive
   / #6.102([uint, [* a]])
 
-redeemer = [ tag: redeemer_tag, index: uint, data: plutus_data, ex_units: ex_units ]
+redeemers =
+  [ + [ tag: redeemer_tag, index: uint, data: plutus_data, ex_units: ex_units ] ]
+; TODO: Add alternative implementation that reflects the reality more accuratly:
```
Comment (Soupstraw) [inline eras/conway/impl/cddl-files/conway.cddl:474]: Make sure there's an issue for this

Comment (lehins) [reply eras/conway/impl/cddl-files/conway.cddl:474]: Thank you for keeping me honest :wink: 
There is already one #3922

File: eras/alonzo/impl/CHANGELOG.md [hunk 2]
```diff
@@ -2,6 +2,18 @@
 
 ## 1.6.0.0
 
+* Deprecated `getAlonzoSpendingTxIn` in favor of `plutusPurposeSpendingTxIn`
+* Remove `rdptr` in favor of new `redeemerPointer`
```
Comment (aniketd) [inline eras/alonzo/impl/CHANGELOG.md]: Reads like a duplicate of line 14.

File: eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs [hunk 3]
```diff
@@ -604,3 +640,71 @@ instance
...
+  forall era.
+  MaryEraTxBody era =>
+  TxBody era ->
+  AlonzoPlutusPurpose AsItem era ->
+  StrictMaybe (AlonzoPlutusPurpose AsIndex era)
+alonzoRedeemerPointer txBody = \case
+  AlonzoSpending txIn ->
+    AlonzoSpending <$> indexOf txIn (txBody ^. inputsTxBodyL)
+  AlonzoMinting policyID ->
+    AlonzoMinting <$> indexOf policyID (txBody ^. mintedTxBodyF :: Set (PolicyID (EraCrypto era)))
```
Comment (Soupstraw) [inline eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs:654]: Should we switch the type of `mintedTxBody` to be an `OSet`?

Comment (lehins) [reply eras/alonzo/impl/src/Cardano/Ledger/Alonzo/TxBody.hs:654]: That would be great, but we can't do that right now, unfortunately
We would have to change `mintTxBodyF` to an `OMap` first.

File: eras/alonzo/impl/CHANGELOG.md (hunk 2 above)
Comment (lehins) [reply eras/alonzo/impl/CHANGELOG.md]: It is indeed a duplicate. Thanks.

**Review by Soupstraw — APPROVED** [review]
LGTM

**Pull request merged**